### PR TITLE
firefox, thunderbird, librewolf: Enable wayland support by default

### LINKED ISSUE #163206: Firefox - consider enabling Wayland by default
### Describe the bug

ref: https://www.phoronix.com/scan.php?page=news_item&px=Firefox-98.0

It sounds like:
- other distros are doing it
- it's about hesitancy around other bits of the user/gpu-space, where we know we're pretty up-to-date

It might also be overkill, we have `firefox-wayland`, but I'm also eager to test out the soon-to-be default path asap - I've been waiting for firefox+wayland for way, way too long.

### PR DESCRIPTION
**Pull request by mweinelt**
Please read the commit message for details.

TL;DR: Always set `MOZ_ENABLE_WAYLAND=1` to prevent XWayland usage on Wayland.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Fixes #163206

### CODE REVIEW

**Review by squalus — APPROVED** [review]

**Comment by bjornfor** [comment]
So using `MOZ_ENABLE_WAYLAND=1` also on X11 is fine?

Does anyone know why firefox doesn't enable this by default itself?

I've run with MOZ_ENABLE_WAYLAND=1 for a while on Wayland, and it does fix the occasionally black screen / crash on startup, so +1 from me.

File: pkgs/applications/networking/browsers/firefox/wrapper.nix [hunk 1]
```diff
@@ -309,7 +308,7 @@ let
             --set MOZ_ALLOW_DOWNGRADE 1 \
             --prefix XDG_DATA_DIRS : "$GSETTINGS_SCHEMAS_PATH" \
             --suffix XDG_DATA_DIRS : '${gnome.adwaita-icon-theme}/share' \
-            ${lib.optionalString forceWayland "--set MOZ_ENABLE_WAYLAND 1"} \
+            --set MOZ_ENABLE_WAYLAND 1
```
Comment (vcunat) [inline pkgs/applications/networking/browsers/firefox/wrapper.nix:311]: Missing backslash.  Without it it won't start.

**Review by vcunat — DISMISSED** [review]

**Comment by QuantMint** [comment]
> So using `MOZ_ENABLE_WAYLAND=1` also on X11 is fine?
> 
> Does anyone know why firefox doesn't enable this by default itself?
> 
> I've run with MOZ_ENABLE_WAYLAND=1 for a while on Wayland, and it does fix the occasionally black screen / crash on startup, so +1 from me.

Some bugs supposedly:
https://bugzilla.mozilla.org/show_bug.cgi?id=1752398

**Comment by vcunat** [comment]
Funnily, there are both kinds in there &ndash; bugs in running on Xwayland ("Blocks") and bugs in running on Wayland ("Depends on").

**Review by lovesegfault — APPROVED** [review]

**Review by Atemu — COMMENTED** [review]
On the one hand, `firefox-wayland` is superior to the experience on xwayland and I've been using it because of that but OTOH Mozilla explicitly decided against enabling wayland by default and it's kind of their decision whether they deem their wayland support to be production-ready.

What Fedora did here was to enable wayland by default in `firefox` but offer a `firefox-x11`. I think we should do the same.

**Comment by lovesegfault** [comment]
Ah, right, this might have licensing implications since we'd be distributing a branded non-standard build.

Sigh.

**Comment by rhendric** [comment]
All this does is set an environment variable in the wrapper script we use to invoke Firefox; does that really constitute changing the build for purposes of licensing?

**Comment by Atemu** [comment]
Well, the thing is, we *could* severely degrade UX by enabling wayland. You and I are comfortable with that but some of our users wouldn't even know that the terrible bug they just experienced could have been caused by wayland being enabled or that that's even the case.

I don't think Mozilla will be coming after us in any capacity if we did a minor thing like this (even if they'd rather we didn't) but, as a packager, it's a general courtesy to follow upstream projects explicit UX decisions wherever possible.  

Fedora stoke a nice balance I think but, from Mozilla's perspective, they'd probably rather have us keep it the way it was (convenient yet explicit wayland opt-in).

**Comment by piegamesde** [comment]
I was baffled to find out that I'd been using Firefox with XWayland all the time. As of 2022, I'd expect applications to use Wayland by default if they support it. Since this is just an environment variable and not a compile flag, and that setting it does not affect X11 support, I am not concerned about licensing stuff.

However, I think we should still provide a way to opt-out of Wayland, i.e. to force the usage of XWayland if desired. Since we are setting the environment variable in the wrapper, setting `MOZ_ENABLE_WAYLAND=0` won't work. Having a `firefox-x11` package as discussed above is a viable option.

An alternative would be to set this environment variable globally, for example as part of a `programs.firefox.wayland` option that defaults to true. The obvious disadvantage is that this would then only apply to NixOS systems.

**Comment by squalus** [comment]
> An alternative would be to set this environment variable globally, for example as part of a `programs.firefox.wayland` option that defaults to true. The obvious disadvantage is that this would then only apply to NixOS systems.

A nixpkgs config option would work, and apply to non-NixOS systems.

**Comment by ajs124** [comment]
> Since we are setting the environment variable in the wrapper, setting `MOZ_ENABLE_WAYLAND=0` won't work

if we use `--set-default` instead of `--set`, this won't be an issue

**Comment by jtojnar** [comment]
> I was baffled to find out that I'd been using Firefox with XWayland all the time. As of 2022, I'd expect applications to use Wayland by default if they support it.

Then that should be taken up with Mozilla.

> Since this is just an environment variable and not a compile flag, and that setting it does not affect X11 support, I am not concerned about licensing stuff.

The method does not really matter. `GTK_USE_PORTAL` was also an environment variable and it lead to a massive slowdown of some users’ boot that was difficult to pinpoint. The issue is that we would be potentially making a non-trivial change in behaviour.

And since Mozilla only allows us to use the Firefox trademark as long as our changes do not break their product, it is our responsibility to make sure that any changes are in fact non-breaking, if we want to keep using the Firefox branding.

At minimum, we should probably find out why exactly it is not enabled by default. But final decision whether the change would be fine with Mozilla can only be done by Nixpkgs Firefox maintainers, hoping that their mental model of Mozilla is good enough.

**Review by piegamesde — APPROVED** [review]

**Comment by zhaofengli** [comment]
> if we use --set-default instead of --set, this won't be an issue

`--set-default` doesn't help because Wayland is used [if `MOZ_ENABLE_WAYLAND` exists at all](https://github.com/mozilla/gecko-dev/blob/8cfe1317c14fe6c997193141b8807d58f92c4f48/toolkit/xre/nsAppRunner.cpp#L4499). In other words, `MOZ_ENABLE_WAYLAND=1` and `MOZ_ENABLE_WAYLAND=0` (or anything else) [do the same thing](https://bugzilla.mozilla.org/show_bug.cgi?id=1588129).

**Comment by colemickens** [comment]
(fwiw, I generally agree with @jtojnar)

**Comment by mweinelt** [comment]
The basic change here is that nobody will use XWayland by default anymore. Nobody will be able to pick the wrong firefox package anymore. XWayland is most certainly the variant that gets the least maintenance at this point.

I think keeping a `firefox-x11` variant around is pretty much pointless, when Firefox with Wayland enabled will gracefully fallback to X11 when the session is based on X11. There is also the possibility to force the usage of X11 by overriding `MOZ_ENABLE_WAYLAND` as outlined by @ajs124.

We're not the first distro to enable Wayland by default, Ubuntu went first.

I do not expect any licensing issue with this.

**Review by vcunat — APPROVED** [review]
I believe this PR is good as it is now.  The variable is there.

I've also been using this PR on pure X11 since the moment it was open.  The wayland variant has been heavily tested by other nixers for a long time, so it's not anything new.

The variable override sounds sufficient at this point, as so far there's no evidence that it would be commonly needed and it works across all the packages using the wrapper.

File: pkgs/top-level/aliases.nix [hunk 2]
```diff
@@ -1462,6 +1465,7 @@ mapAliases ({
   tftp_hpa = throw "'tftp_hpa' has been renamed to/replaced by 'tftp-hpa'"; # Converted to throw 2022-02-22
   thunderbird-68 = throw "Thunderbird 68 reached end of life with its final release 68.12.0 on 2020-08-25";
   thunderbird-bin-68 = thunderbird-68;
+  thunderbird-wayland = thunderbird;
```
Comment (mweinelt) [inline pkgs/top-level/aliases.nix:1468]:
```suggestion
  thunderbird-wayland = thunderbird; # Added 2022-11-15
```

pushing an update in a second.

**Review by maralorn — APPROVED** [review]
Have been using this for years, works great.

**Review by SamLukeYes — APPROVED** [review]
